Birthday-2024-Project/Scripts/GridLogic.gd

```diff
@@ -1,89 +1,146 @@
-			#instantiate scene piece
-			var scenePiece = scenePiecePrefab.instantiate()
-			add_child(scenePiece)
-			var pieceLogic : PieceLogic = scenePiece as PieceLogic
-			pieceLogic.levelGridReference = self
-			pieceLogic.game_manager = gameManager
-			pieceLogic._initialize()
+			var pieceLogic : PieceLogic = LoadPiece(pieceSetup.pieceID)
 			pieceLogic.SetPieceRotation(pieceSetup.pieceRotation)
-			if pieceSetup.isBlocker:
+			pieceLogic.RecordGridPosition(pieceSetup.gridPosition)
+			if pieceLogic.isBlocker:
 				#a locked piece that sits on the grid
 				SetGridSpacesByPieceShape(pieceSetup.gridPosition, GridSpaceInfo.GridSpaceStatus.CLOSED, pieceLogic)
-				blockerPieces.push_back(scenePiece)
+				pieceLogic.return_piece(true)
 			else:
 				#open level grid spaces
 				SetGridSpacesByPieceShape(pieceSetup.gridPosition, GridSpaceInfo.GridSpaceStatus.OPEN, pieceLogic)
-				availablePieces.push_back(scenePiece)
 		
 		#center grid map based on the width and height
 		var midPoint : Vector2i = _GridCoordinateToPosition(Vector2i(xMaxGrid, yMaxGrid)) + _GridCoordinateToPosition(Vector2i(xMinGrid, yMinGrid))
 		global_position = midPoint * -0.5 + gridCenterOffset
 		
-		#set piece to outskirts of level grid and unrotate
-		#TODO: place more neatly around the grid
-		var iter = 0
-		for availablePiece in availablePieces:
-			availablePiece.global_position = GridCoordinateToPosition(startingPositions[iter]) #shunting it to the right side
-			(availablePiece as PieceLogic).SetPieceRotation(PieceLogic.RotationStates.DEG_0)
-			(availablePiece as PieceLogic)._SetReturnPoint()
-			iter += 1
+		if(gridMode == GridMode.PLAY):
+			#set piece to outskirts of level grid and unrotate
+			#TODO: place more neatly around the grid
+			var iter = 0
+			for availablePiece in availablePieces:
+				availablePiece.global_position = GridCoordinateToPosition(startingPositions[iter]) #shunting it to the right side
+				availablePiece.SetPieceRotation(PieceLogic.RotationStates.DEG_0)
+				availablePiece._SetReturnPoint()
+				iter += 1
+		else:
+			global_position = gridCenterOffset
+			for availablePiece in availablePieces:
+				availablePiece.return_piece(true)
+
+func LoadPiece(pieceID : String) -> PieceLogic:
+	#look up scene piece via ID
+	var scenePiecePrefab = load("res://ScenePrefabs/Pieces/" + pieceID + ".tscn")
+	#instantiate scene piece
+	var scenePiece = scenePiecePrefab.instantiate()
+	add_child(scenePiece)
+	var pieceLogic : PieceLogic = scenePiece as PieceLogic
+	pieceLogic.levelGridReference = self
+	pieceLogic.game_manager = gameManager
+	pieceLogic._initialize()
+	pieceLogic.piece_id = pieceID
+	pieceLogic.SetPieceRotation(PieceLogic.RotationStates.DEG_0)
+	if pieceLogic.isBlocker:
+		blockerPieces.push_back(pieceLogic)
+	else:
+		availablePieces.push_back(pieceLogic)
+	return pieceLogic
+
+func DeletePiece(piece : PieceLogic):
+	availablePieces.erase(piece)
+	blockerPieces.erase(piece)
+	piece.queue_free()
+
+func ExportLevel() -> String:
```

Comment:
general idea: exporting levels to JSON files that can be later used during level selection

Reply:
I'm touching upon this because it would be easier to have it in file/files with incoming submission phase. Or at least I think so

Reply:
I think it's more flexible if we do it on our end. Technically, the LevelSetup are the files, but they aren't very clearly defined since we haven't talked about it much. As you can see above, I left an art file field, but it seems like we won't be using that, and I forgot to add an author name. We can update the LevelSetup as we wish if we keep them on our end. Another thing we can add could be an @export reference to the level preview image for level select

Secondly, there would be a some UI and technical overhead we'd have to deal with. Such as, have the user provide an output folder for the file. And file management, which I'm not very familiar with in Godot.

Lastly, though this is just my opinion, but I think it's easier for submitters if it's text to paste. It goes from 2 steps (generate file, submit file) to 1 step (paste json text). Some might not like creating a file on their local machine (this could all be in browser). Everyone knows how to paste. It's all minor stuff, but anything that can improve accessibility means more people can participate. Though, contributing an entire level is already a pretty sizeable ask